File: scripts/__pbot/PBotInventory.py
from typing import List, Optional, Tuple
from .PBotItem import PBotItem

class PBotInventory(object):
    def __init__(self, inv):
        self.__inv = inv

    ## Return all items that the inventory contains
    # @return list of items in the inventory
    def get_inventory_items(self) -> List[PBotItem]:
        return [PBotItem(x) for x in self.__inv.getInventoryContents()]

    ## Returns a list of items with name matching regex pattern from the inventory
    # @param pattern Regex pattern matching item names
    # @return List of items with name matching given pattern
    def get_inventory_items_by_names(self, pattern: str) -> List[PBotItem]:
        return [PBotItem(x) for x in self.__inv.getInventoryItemsByNames(pattern)]

    ## Returns a list of items with resname matching regex pattern from the inventory
    # @param pattern Regex pattern matching item resnames
    # @return List of items with resname matching pattern
    def get_inventory_items_by_resnames(self, pattern: str) -> List[PBotItem]:
        return [PBotItem(x) for x in self.__inv.getInventoryItemsByResnames(pattern)]

    ## Finds the item with certain location from the inventory
    # @param x x-coordinate of the item location in inventory
    # @param y y-coordinate of the item location in inventory
    # @return None if not found
    def get_item_by_location(self, x: int, y: int) -> Optional[PBotItem]:
        itm = self.__inv.getItemFromInventoryAtLocation(x, y)
        return PBotItem(itm) if itm is not None else None

    ## Drop item from the hand to given slot in the inventory
    # @param x x-coordinate in inventory to drop the item into
    # @param y y-coordinate in inventory to drop the item into
    def drop_item_to_inventory(self, x: int, y: int):
        self.__inv.dropItemToInventory(x, y)

    ## Amount of free slots in the inventory
    # @return Amount of free inventory slots
    def free_slots(self) -> int:
        return self.__inv.freeSlotsInv()

    ## Transfer 1 item by scrolling
    def xfer_to(self):
        self.__inv.xferTo()

    ## Size of the inventory slots
    # @return x y size
    def size(self) -> Tuple[int, int]:
        sz = self.__inv.size()
        return sz.x, sz.y,

## Inventory where the item is located in
# @return inventory or None if not found
def get_inventory_for_item(item: PBotItem) -> Optional["PBotInventory"]:
    inv = item._item.getInventory()
    return PBotInventory(inv) if inv is not None else None

File: src/haven/purus/pbot/api/PBotInventory.java
package haven.purus.pbot.api;

import haven.Coord;
import haven.Inventory;
import haven.UI;
import haven.WItem;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class PBotInventory {

	private final Inventory inventory;
	private final PBotSession pBotSession;

	PBotInventory(Inventory inventory, PBotSession pBotSession) {
		this.inventory = inventory;
		this.pBotSession = pBotSession;
	}

	/**
	 * Return all items that the inventory contains
	 * @return List of items in the inventory
	 */
	public List<PBotItem> getInventoryContents() {
		return inventory.children(WItem.class).stream()
				.map(witem -> new PBotItem(witem.item, pBotSession))
				.collect(Collectors.toList());
	}

	/**
	 * Returns a list of items with specific regex pattern from the inventory
	 * @param pattern Regex pattern matching item names
	 * @return List of items with name matching given pattern
	 */
	public List<PBotItem> getInventoryItemsByNames(String pattern) {
		Pattern pat = Pattern.compile(pattern);
		return inventory.children(WItem.class).stream()
				.map(witem -> new PBotItem(witem.item, pBotSession))
				.filter(item -> {
					String name = item.getName();
					return (name != null && pat.matcher(name).matches());
				})
				.collect(Collectors.toList());
	}

	/**
	 * @param pattern Regex pattern matching item resnames
	 * @return List of items with resname matching pattern
	 */
	public List<PBotItem> getInventoryItemsByResnames(String pattern) {
		List<PBotItem> items = new ArrayList<>();
		Pattern pat = Pattern.compile(pattern);
		return inventory.children(WItem.class).stream()
				.map(witem -> new PBotItem(witem.item, pBotSession))
				.filter(item -> {
					String resname = item.getResname();
					return (resname != null && pat.matcher(resname).matches());
				})
				.collect(Collectors.toList());
	}

	/**
	 * Finds an item with certain location from the inventory
	 * @param x x-coordinate of the item location in inventory
	 * @param y y-coordinate of the item location in inventory
	 * @return Null if not found
	 */
	public PBotItem getItemFromInventoryAtLocation(int x, int y) {
		for(WItem witem : inventory.children(WItem.class)) {
			if(witem.c.div(UI.scale(33)).x == x && witem.c.div(UI.scale(33)).y == y) {
				return new PBotItem(witem.item, pBotSession);
			}
		}
		return null;
	}

	/**
	 * Drop item from the hand to given slot in inventory
	 * @param x x coordinate in inventory to drop the item into
	 * @param y y coordinate in inventory to drop the item into
	 */
	public void dropItemToInventory(int x, int y) {
		inventory.wdgmsg("drop", new Coord(x, y));
	}

	/**
	 * Amount of free slots in the inventory
	 * @return Amount of free inventory slots
	 */
	public int freeSlotsInv() {
		return inventory.isz.x * inventory.isz.y -  inventory.children(WItem.class)
				.stream().map(witem -> witem.sz.div(UI.scale(33)))
				.mapToInt((c) -> c.x * c.y)
				.sum();
	}

	/**
	 * Transfer 1 item by scrolling
	 */
	public void xferTo() {
		pBotSession.PBotUtils().playerInventory().inventory.wdgmsg("invxf", inventory.wdgid(), 1);
	}

	/**
	 * Size of the inventory
	 * @return Size
	 */
	public Coord size() {
		return this.inventory.isz;
	}
}
